server/athenian/api/controllers/search_controller/search_prs.py

```diff
@@ -220,3 +240,137 @@ async def _search_pr_digests(
             ",".join(map(str, unknown_prs)),
         )
     return pr_digests
+
+
+@sentry_span
+def _apply_order_by(
+    pr_facts: pd.DataFrame,
+    search_filter: _SearchPRsFilter,
+    order_by: Sequence[SearchPullRequestsOrderByExpression],
+) -> pd.DataFrame:
+    assert order_by
+    if not len(pr_facts):
+        return pr_facts
+
+    if any(expr.field in _OrderByMetrics.FIELDS for expr in order_by):
+        order_by_metrics = _OrderByMetrics.build(pr_facts, search_filter, order_by)
+    else:
+        order_by_metrics = None
+
+    order_keys: list[npt.NDArray] = []
+    keep_mask = np.full((len(pr_facts),), True, bool)
+
+    for expr in order_by:
+        if expr.field in _OrderByMetrics.FIELDS:
+            assert order_by_metrics is not None
+            exp_order_keys, discard = order_by_metrics.parse_expression(expr)
+            order_keys.append(exp_order_keys)
+            keep_mask[discard] = False
+        else:
+            raise ValueError(f"Invalid order by field {expr.field}")
+
+    sorted_positions = np.lexsort(list(reversed(order_keys)))
+    kept_positions = sorted_positions[np.nonzero(keep_mask[sorted_positions])]
+    return pr_facts.iloc[kept_positions]
+
+
+class _OrderByMetrics:
+    """Handles order by pull request metric values.
+
+    Expressions about a metric field can be fed into this object with `parse_expression`.
+    """
+
+    FIELDS = pr_metric_calculators
+
+    def __init__(self, calc_ensemble: PullRequestMetricCalculatorEnsemble):
+        self._calc_ensemble = calc_ensemble
+
+    @classmethod
+    def build(
+        cls,
+        pr_facts: pd.DataFrame,
+        search_filter: _SearchPRsFilter,
+        order_by: Sequence[SearchPullRequestsOrderByExpression],
+    ) -> Optional[_OrderByMetrics]:
+        metrics = [expr.field for expr in order_by if expr.field in cls.FIELDS]
+        assert metrics
+        min_times = np.array(
+            [search_filter.time_from.replace(tzinfo=None)], dtype="datetime64[ns]",
+        )
+        max_times = np.array([search_filter.time_to.replace(tzinfo=None)], dtype="datetime64[ns]")
+        calc_ensemble = PullRequestMetricCalculatorEnsemble(
+            *metrics, quantiles=(0, 1), quantile_stride=DEFAULT_QUANTILE_STRIDE,
+        )
+        groups = np.full((1, len(pr_facts)), True, bool)
+        calc_ensemble(pr_facts, min_times, max_times, groups)
+        return cls(calc_ensemble)
+
+    def parse_expression(
+        self,
+        expr: SearchPullRequestsOrderByExpression,
+    ) -> tuple[npt.NDArray, npt.NDArray[int]]:
+        """Parse an expression and return a tuple with order keys and discard indexes.
```

Comment:
which expression? not clear...

Reply:
the `SearchPullRequestsOrderByExpression` passed as argument, added "order by expression" to be more clear